src/kakarot/instructions.cairo

```diff
@@ -34,80 +34,579 @@ from kakarot.instructions.sha3 import Sha3
 // @custom:namespace EVMInstructions
 namespace EVMInstructions {
     // @notice Decode the current opcode and execute associated function.
-    // @dev The function iterates through the provided instructions and executes each of them
-    //      whilst also performing safety checks and updating the pc counter after each instruction execution
-    // @param instructions The instruction set.
+    // @dev The function uses an internal jump table to execute the corresponding opcode
     // @param ctx The pointer to the execution context.
     // @return The pointer to the updated execution context.
-
     func decode_and_execute{
         syscall_ptr: felt*,
         pedersen_ptr: HashBuiltin*,
-        range_check_ptr,
+        range_check_ptr: felt,
         bitwise_ptr: BitwiseBuiltin*,
-    }(instructions: felt*, ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
         alloc_locals;
 
         // Retrieve the current program counter.
         let pc = ctx.program_counter;
-
-        // Revert if pc < 0
-        with_attr error_message("Kakarot: InvalidCodeOffset") {
-            assert_nn(pc);
-        }
-
         local opcode;
 
-        let is_pc_le_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
+        let is_pc_ge_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
 
-        if (is_pc_le_code_len == 1) {
+        if (is_pc_ge_code_len == TRUE) {
             assert opcode = 0;
         } else {
             assert opcode = [ctx.call_context.bytecode + pc];
         }
+        tempvar offset = 1 + 4 * opcode;
 
         // move program counter + 1 after opcode is read
         let ctx = ExecutionContext.increment_program_counter(self=ctx, inc_value=1);
 
-        // Revert if opcode does not exist
-        with_attr error_message("Kakarot: UnknownOpcode {opcode}") {
-            // Read opcode in instruction set
-            let function_codeoffset_felt = instructions[opcode];
-            let function_codeoffset = cast(function_codeoffset_felt, codeoffset);
-            let (function_ptr) = get_label_location(function_codeoffset);
-        }
-
-        // Prepare implicit arguments
-        let implicit_args_len = decode_and_execute.ImplicitArgs.SIZE;
-        tempvar implicit_args = new decode_and_execute.ImplicitArgs(syscall_ptr, pedersen_ptr, range_check_ptr, bitwise_ptr);
-
-        // Build arguments array
-        let (args_len: felt, args: felt*) = prepare_arguments(
-            ctx=ctx, implicit_args_len=implicit_args_len, implicit_args=implicit_args
-        );
-
-        // Invoke opcode function
-        invoke(function_ptr, args_len, args);
+        // Prepare arguments
+        [ap] = syscall_ptr, ap++;
+        [ap] = pedersen_ptr, ap++;
+        [ap] = range_check_ptr, ap++;
+        [ap] = bitwise_ptr, ap++;
+        [ap] = ctx, ap++;
+
+        // call opcode
+        jmp rel offset;
+        call exec_stop;
```

Comment:
can you add a comment to state this is opcode number 0? 
when is this called? when offset == 0?